sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/client.py

```diff
@@ -409,16 +440,16 @@ def mgmt_request(self, message, **kwargs):
         operation_type = kwargs.pop("operation_type", None)
         node = kwargs.pop("node", "$management")
         timeout = kwargs.pop("timeout", 0)
-        try:
-            mgmt_link = self._mgmt_links[node]
-        except KeyError:
-            mgmt_link = ManagementOperation(self._session, endpoint=node, **kwargs)
-            self._mgmt_links[node] = mgmt_link
-            mgmt_link.open()
-
-            while not mgmt_link.ready():
-                self._connection.listen(wait=False)
-
+        with self._lock:
```

Comment:
I know you've discussed it in our sync-ups - but can you expand on what you were observing that made this necessary?
I'm trying to figure out where the race condition was that we need to actually lock - is it `self._mgmt_links`? Or something on the connection?

Reply:
So the test where I saw this threading context switching was in (sync) test_queue_by_queue_client_conn_str_receive_handler_with_autolockrenew where we set AutoLockRenewer max_workers to be 8. 

The flow of the autolockrenewing (at least what I have seen so far):

In AutoLockRenewer:   

The Main Thread initializes a dispatch worker in _init_workers(). In dispatch_worker() since we have more than one worker we use the ThreadPoolExecutor to create tasks for _auto_lock_renew_task(). If the message needs to be renewed we call renew_message_lock() with one of the threads created by the ThreadPoolExecutor. In this specific example 2 threads (call them 1 and 2) enter into this function one after another. 

In the Receiver:

Entering into the SB reciever, renew_message_lock() calls self._renew_locks(). These 2 threads, 1 and 2, enter this function one after another as well. They then both enter into self._mgmt_request_response_with_retry() which calls client.py.

In pyamqp:

In client.py mgmt_request, these two threads both enter one after another (print statements back to back so context switching). Since thread 1 and thread 2 enter back to back, Thread 1 hits a KeyError and tries to create a new mgmt_link. Switching to Thread 2, it also hits the KeyError as Thread1 has not had time to open() the link yet. For both of these we wait for the client.ready() on the mgmt_link and sleep for .5 seconds if it isn't ready.


(In print Statments) 

More threads go through this process all now using existing mgmt_link (but there are 2). Then you hit this error:

`Unexpected error occurred (ValueError('Connection not open.')). Handler shutting down.`
`AttributeError: 'ServiceBusReceivedMessage' object has no attribute 'renew_lock'`
`Connection closed with error: [b'amqp:not-found', b"The session channel '1' cannot be found in connection 'Connection'.", None]
Unexpected error occurred (ValueError('Management Operation error occurred.')). Handler shutting down.`

and also in the stacktrace: 

` while not self._handler.client_ready():
AttributeError: 'NoneType' object has no attribute 'client_ready'`

So then I went and compared this against uamqp. While pyamqp initially creates 2 mgmt_links because the one thread doesn't recognize the other mgmt_link being made in time, uamqp uses the same mgmt link for all renew lock operations. By using the lock in pyamqp, we ended up with one exisitng mgmt_link.

Reply:
This was very helpful - thank you!
While it's possible we could do the locking somewhere else - I think this is a good first step towards making the clients more threadsafe.